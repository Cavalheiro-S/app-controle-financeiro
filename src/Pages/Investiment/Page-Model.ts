import moment from 'moment';

class PageModel {

    private name : String;
    private type : String | null;
    private date : String;
    private value : Number;

    constructor(idName: string, idType: string, idValue: string, idDate: string){
        this.name = document.querySelector<HTMLInputElement>(`#${idName}`)!.value
        this.type = document.querySelector<HTMLInputElement>(`#${idType}`)!.value
        this.value = document.querySelector<HTMLInputElement>(`#${idValue}`)!.valueAsNumber
        this.date = moment(document.querySelector<HTMLInputElement>(`#${idDate}`)!.valueAsDate).add(1,"d").format("YYYY-MM-DD");
    }

    makeObjectWithValues() : Object{
        console.log(this.name, this.type, this.date, this.value);
        
        return{
            name: this.name,
            type: this.type,
            value: this.value,
            date: this.date
        }
    }
    
}

export default PageModel