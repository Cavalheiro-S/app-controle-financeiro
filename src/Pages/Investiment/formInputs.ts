import { TypeEnum } from "../../common/enum/typeEnum"
import { InputProps } from "../../common/interface/InputProps"

export const inputsInvestiment: InputProps[] = [
    {
        id: "inputNameInvestiment",
        type: TypeEnum.text,
        placeholder: "Nome do Investimento"
    },
    {
        id: "inputTypeInvestiment",
        type: TypeEnum.text,
        placeholder: "Tipo do Investimento",
        select: {
            isSelect: true,
            optionsOfSelect: ["Renda Vari√°vel", "Renda Fixa", "Outro"]
        }
    },
    {
        id: "inputDateInvestiment",
        type: TypeEnum.date,
        placeholder: "Data do Investimento"
    },
    {
        id: "inputValueInvestiment",
        type: TypeEnum.number,
        placeholder: "Valor do Investimento"
    }
]