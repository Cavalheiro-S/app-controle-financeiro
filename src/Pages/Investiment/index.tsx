import { Card } from "../../components/Card";
import Table from "../../components/Table";
import { ObjectTable } from "../../common/interface";
import { useEffect, useState } from "react";
import axios from "axios";
import { Button, MenuItem, Select, TextField } from "@mui/material";
import PageModel from "./Page-Model";
import moment from "moment";

const Investiment = () => {

    const [itemsOfTable, setItemsOfTable] = useState<ObjectTable[] | null>();

    function updateTable() {
        axios.get<ObjectTable[]>("http://localhost:4000/investiment/")
            .then(tableData => setItemsOfTable(tableData.data));
    }

    useEffect(() => {
        updateTable();
    }, [])

    function AddNewInvestiment(event:React.FormEvent): void {
        event.preventDefault();
        const investimentModel = new PageModel("inputNameInvestiment","inputTypeInvestiment","inputValueInvestiment", "inputDateInvestiment");
        axios.post("http://localhost:4000/investiment/", investimentModel.makeObjectWithValues())
            .then(() => updateTable());

    }

    function removeInvestiment(id: number) {
        axios.delete(`http://localhost:4000/investiment/${id}`)
            .then(() => updateTable());
    }

    return (
        <section className="container__page container__page--flex">

            <Card
                firstLineCard="Adicionando Um"
                title="Investimento"
                describe="Preencha as informações para adicionar uma nova investimento"
            >
                <form className="form" onSubmit={event => AddNewInvestiment(event)}>
                    <TextField required fullWidth id="inputNameInvestiment" label="Nome do Investimento" />
                    <Select defaultValue="Renda Variavel" fullWidth id="inputTypeInvestiment">
                        <MenuItem value="Renda Variavel">Renda Variavel</MenuItem>
                        <MenuItem value="Renda Fixa">Renda Fixa</MenuItem>
                        <MenuItem value="outros">Outros</MenuItem>
                    </Select>
                    <TextField required fullWidth id="inputDateInvestiment" type="date"/>
                    <TextField required fullWidth id="inputValueInvestiment" type="number" label="Valor do Investimento" />
                    <Button type="submit" classes={{ root: "button" }} fullWidth id="buttonInvestiment">Adicionar </Button>
                </form>

            </Card>
            <Table
                id="tableInvestiment"
                removeItemTable={removeInvestiment}
                tableTitle={["Investimentos", " Adicionados"]}
                titleHead={["Nome", "Tipo", "Data", "Valor"]}
                tableData={itemsOfTable} />

        </section>

    )

}

export default Investiment;
