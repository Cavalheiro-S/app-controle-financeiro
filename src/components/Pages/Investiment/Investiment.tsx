import { Card } from "../../Card";
import { Table } from "../../Table";
import { ObjectTable } from "../../../common/interfaces/TableProps";
import { useContext } from "react";
import { Form } from "../../Form";
import { TypeProps } from "../../../common/interfaces/FormProps";
import { ButtonProps } from "../../../common/interfaces/ButtonProps";
import { InputProps } from "../../../common/interfaces/InputProps";
import { InvestimentContext } from "../../../common/context/InvestimentContext";

const Investiment = () => {

    const investimentContext = useContext(InvestimentContext);
    function buttonClicked(): void {
        const name = document.querySelector("#inputNameInvestiment") as HTMLInputElement;
        const type = document.querySelector("#inputTypeInvestiment") as HTMLInputElement;
        const value = document.querySelector("#inputValueInvestiment") as HTMLInputElement;

        if (value.value === "") {
            value.focus();
            alert("O valor não pode ser deixado vazio");
        }
        else {
            const objTableData: ObjectTable = {
                id: investimentContext!.idObjectTable,
                name: name.value,
                type: type.value,
                value: value.valueAsNumber
            }

            investimentContext?.setObjectsTable([...investimentContext.objectsTable, objTableData]);
            investimentContext?.setIdObjectTable(investimentContext.idObjectTable + 1);
            investimentContext?.setValueTotalInvestiment(investimentContext.valueTotalInvestiment + objTableData.value);
        }

    }

    function removeItemFromTable(id: number) {

        const newArrayOfObjects = investimentContext?.objectsTable.filter(
            objectTable => objectTable.id !== id ? objectTable : false
        );

        investimentContext?.setObjectsTable(newArrayOfObjects as ObjectTable[]);
    }

    const inputsTemp: InputProps[] = [
        {
            id: "inputNameInvestiment",
            type: TypeProps.text,
            placeholder: "Nome do Investimento"
        },
        {
            id: "inputTypeInvestiment",
            type: TypeProps.text,
            placeholder: "Tipo do Investimento",
            select: {
                isSelect: true,
                optionsOfSelect: ["Renda Variável", "Renda Fixa", "Outro"]
            }
        },
        {
            id: "inputValueInvestiment",
            type: TypeProps.number,
            placeholder: "Valor do Investimento"
        }
    ]

    const buttonsTemp: ButtonProps[] = [
        {
            text: "Adicionar",
            buttonClickedFunction: buttonClicked
        }]

    return (
        <section className="container__page container__page--flex">

            <Table
                removeItemTable={removeItemFromTable}
                tableTitle={["Investimentos", " Adicionados"]}
                titleHead={["Nome", "Tipo", "Valor"]}
                tableData={investimentContext?.objectsTable} />
            <Card
                firstLineCard="Adicionando Um"
                title="Investimento"
                describe="Preencha as informações para adicionar uma nova investimento"
            >
                <Form inputs={inputsTemp} buttons={buttonsTemp} />
            </Card>
        </section>

    )

}

export default Investiment;