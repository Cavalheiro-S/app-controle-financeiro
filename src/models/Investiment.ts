import moment from "moment";
import { ObjectTable } from "../common/interfaces/TableProps";
export class Investiment {
    
    private _name:string ;
    private _type:string;
    private _value:number;
    private _date:string;
    

    constructor(nameID: string, typeID: string, valueID: string, dateID: string ){
        const inputName = document.querySelector(`#${nameID}`) as HTMLInputElement
        const inputType = document.querySelector(`#${typeID}`) as HTMLInputElement
        const inputValue = document.querySelector(`#${valueID}`) as HTMLInputElement
        const inputDate = document.querySelector(`#${dateID}`) as HTMLInputElement

        this._name = inputName.value
        this._type = inputType.value
        this._value = inputValue.valueAsNumber
        this._date = inputDate.value 
        ? moment().add(1,"d").format("YYYY-MM-DD") 
        : moment(inputDate.valueAsDate).add(1,"d").format("YYYY-MM-DD");
    }

    public get name(){
        return this._name
    }

    public get type(){
        return this._type
    }

    public get value(){
        return this._value
    }

    public get date(){
        return this._date
    }


    public makeObject() : ObjectTable{

        return{
            id: Math.floor(Math.random() * 100),
            name:this._name,
            type:this._type,
            value:this._value,
            date:this._date
        }
    }
}