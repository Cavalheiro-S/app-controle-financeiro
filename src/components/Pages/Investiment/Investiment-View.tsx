import { Card } from "../../Card";
import { Table } from "../../Table";
import { ObjectTable } from "../../../common/interfaces/TableProps";
import { SetStateAction, useContext, useEffect, useMemo, useState } from "react";
import { Form } from "../../Form";
import { TypeProps } from "../../../common/interfaces/FormProps";
import { ButtonProps } from "../../../common/interfaces/ButtonProps";
import { InputProps } from "../../../common/interfaces/InputProps";
import { InvestimentContext } from "../../../common/context/InvestimentContext";
import axios from "axios";
import { Investiment } from "../../../models/Investiment";


const InvestimentView = () => {
    const [itemsOfTable, setItemsOfTable] = useState<ObjectTable[] | null>();
    

    useEffect( () => {
        axios.get<ObjectTable[]>("http://localhost:5000/investiment/all")
        .then( table => {
            setItemsOfTable(table.data)
        })

    })

    function buttonClicked(): void {

        const investiment = new Investiment(
            "inputNameInvestiment",
            "inputTypeInvestiment",
            "inputValueInvestiment",
            "inputDateInvestiment");

        axios.post("http://localhost:5000/investiment/add", {
            name: investiment.name,
            type: investiment.type,
            value: investiment.value,
            date: investiment.date
        }).then(resp => console.log(resp.data))


    }

    function removeItemFromTable(id: number) {
        axios.post("http://localhost:5000/investiment/delete", {
            id
        }).then(() => console.log("Item retirado"))
    }

    const inputsTemp: InputProps[] = [
        {
            id: "inputNameInvestiment",
            type: TypeProps.text,
            placeholder: "Nome do Investimento"
        },
        {
            id: "inputTypeInvestiment",
            type: TypeProps.text,
            placeholder: "Tipo do Investimento",
            select: {
                isSelect: true,
                optionsOfSelect: ["Renda Variável", "Renda Fixa", "Outro"]
            }
        },
        {
            id: "inputDateInvestiment",
            type: TypeProps.date,
            placeholder: "Data do Investimento"
        },
        {
            id: "inputValueInvestiment",
            type: TypeProps.number,
            placeholder: "Valor do Investimento"
        }
    ]

    const buttonsTemp: ButtonProps[] = [
        {
            text: "Adicionar",
            buttonClickedFunction: buttonClicked
        }]

    return (
        <section className="container__page container__page--flex">

            <Table
                removeItemTable={removeItemFromTable}
                tableTitle={["Investimentos", " Adicionados"]}
                titleHead={["Nome", "Tipo", "Data", "Valor"]}
                tableData={itemsOfTable} />
            <Card
                firstLineCard="Adicionando Um"
                title="Investimento"
                describe="Preencha as informações para adicionar uma nova investimento"
            >
                <Form inputs={inputsTemp} buttons={buttonsTemp} />
            </Card>
        </section>

    )

}

export default InvestimentView;