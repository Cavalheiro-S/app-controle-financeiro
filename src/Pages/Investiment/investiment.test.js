import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Investiment from ".";
import axios from 'axios';
import { act, } from "react-dom/test-utils";
import TestRenderer from "react-test-renderer";


describe("Screen of Investiment", () => {

    test("Test if 'Investiment Card' has renderer", async () => {

       expect(TestRenderer.create(<Investiment/>).toJSON()).toMatchSnapshot();
    })

})