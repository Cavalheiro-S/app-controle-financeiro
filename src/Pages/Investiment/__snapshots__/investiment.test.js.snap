// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen of Investiment Test if 'Investiment Card' has renderer 1`] = `
<section
  className="container__page container__page--flex"
>
  <div
    className="card"
  >
    <h2
      className="card__title"
    >
      Adicionando Um
      <br />
      <span
        className="card__type"
      >
        Investimento
      </span>
    </h2>
    <h3
      className="card__describe"
    >
      Preencha as informações para adicionar uma nova investimento
    </h3>
    <form
      className="form"
    >
      <input
        className="input"
        data-testid="inputNameInvestiment"
        id="inputNameInvestiment"
        placeholder="Nome do Investimento"
        type="text"
      />
      <select
        className="input"
        data-testid="inputTypeInvestiment"
        id="inputTypeInvestiment"
        placeholder="Tipo do Investimento"
      >
        <option
          value="Renda Variável"
        >
          Renda Variável
        </option>
        <option
          value="Renda Fixa"
        >
          Renda Fixa
        </option>
        <option
          value="Outro"
        >
          Outro
        </option>
      </select>
      <input
        className="input"
        data-testid="inputDateInvestiment"
        id="inputDateInvestiment"
        placeholder="Data do Investimento"
        type="date"
      />
      <input
        className="input"
        data-testid="inputValueInvestiment"
        id="inputValueInvestiment"
        placeholder="Valor do Investimento"
        type="number"
      />
      <button
        className="button button--primary"
        data-testid="buttonInvestiment"
        id="buttonInvestiment"
        onClick={[Function]}
        type="submit"
      >
        Adicionar
      </button>
    </form>
  </div>
  <section
    className="section__table "
  >
    <h2
      className="section__title"
    >
      <span
        className="section__span--highlight"
      >
        Investimentos
      </span>
      <span>
         Adicionados
      </span>
    </h2>
    <table
      className="table"
      data-testid="tableInvestiment"
      id="tableInvestiment"
    >
      <thead
        className="thead"
      >
        <tr
          className="table__line"
        >
          <td
            className="table__data table__header"
          >
            Nome
          </td>
          <td
            className="table__data table__header"
          >
            Tipo
          </td>
          <td
            className="table__data table__header"
          >
            Data
          </td>
          <td
            className="table__data table__header"
          >
            Valor
          </td>
        </tr>
      </thead>
      <tbody
        className="tbody"
      />
    </table>
  </section>
</section>
`;
